packages/fxa-auth-server/lib/payments/iap/apple-app-store/purchase-manager.ts

```diff
@@ -0,0 +1,289 @@
+        throw libraryError;
+      }
+    }
+    // Unlike Google Play subscriptions, We don't need to check if the purchase is
+    // registerable (i.e. not terminal), since the original transaction ID is
+    // the primary key in Firestore, and it can be "revived at any time if the customer
+    // returns to any SKU in that subscription group".
+    // See https://developer.apple.com/forums/thread/704167.
+
+    // STEP 2. Check if the purchase has been registered to a user. If it is, then return conflict error to our caller.
+    if (purchase.userId === userId) {
+      // Purchase record already registered to the target user. We'll do nothing.
+      return purchase;
+    } else if (purchase.userId) {
+      this.log.info('Purchase already registered', { purchase });
+      // Purchase record already registered to different user. Return 'conflict' to caller
+      const libraryError = new Error(
+        'Purchase has been registered to another user'
+      );
+      libraryError.name = PurchaseUpdateError.CONFLICT;
+      throw libraryError;
+    }
+
+    // STEP 3: Register purchase to the user
+    await this.forceRegisterToUserAccount(originalTransactionId, userId);
+
+    return purchase;
+  }
+
+  /*
+   * Query subscriptions registered to a particular user that are active.
+   * Note: Other subscriptions which don't meet the above criteria still exists in Firestore purchase records, but are not accessible from outside of the library.
+   */
+  async queryCurrentSubscriptionPurchases(
+    userId: string,
+    bundleId?: string,
+    productId?: string
+  ): Promise<Array<SubscriptionPurchase>> {
+    const purchaseList = new Array<SubscriptionPurchase>();
+
+    try {
+      // Create query to fetch possibly active subscriptions from Firestore
+      let query = this.purchasesDbRef
+        .where('formOfPayment', '==', APPLE_APP_STORE_FORM_OF_PAYMENT)
+        .where('type', '==', TransactionType.AutoRenewableSubscription)
+        .where('userId', '==', userId);
+
+      if (productId) {
+        query = query.where('productId', '==', productId);
+      }
+
+      if (bundleId) {
+        query = query.where('bundleId', '==', bundleId);
+      }
+
+      // Do fetch possibly active subscription from Firestore
+      const queryResult = await query.get();
+
+      // Loop through these subscriptions and filter those that are indeed active
+      for (const purchaseRecordSnapshot of queryResult.docs) {
+        let purchase: SubscriptionPurchase =
+          SubscriptionPurchase.fromFirestoreObject(
+            purchaseRecordSnapshot.data()
+          );
+
+        if (!purchase.isEntitlementActive()) {
+          // If a subscription purchase record in Firestore indicates says that it has expired,
+          // and we know that its status could have been changed since we last fetch its details,
+          // then we should query the App Store Server API to get its latest status
```

Comment:
Why does the entitlement not being active mean that it could've changed?

Reply:
We discussed this sync; I updated the comment to be more clear about why we are doing this.